beacon-chain/state/stategen/replayer.go

```diff
@@ -0,0 +1,347 @@
+	// slots via process_slots.
+	ForSlot(target types.Slot) Replayer
+}
+
+// NewCanonicalBuilder handles initializing the default concrete ReplayerBuilder implementation.
+func NewCanonicalBuilder(h HistoryAccessor, c CanonicalChecker, cs CurrentSlotter) *CanonicalBuilder {
+	return &CanonicalBuilder{
+		chainer: &canonicalChainer{
+			h:  h,
+			c:  c,
+			cs: cs,
+		},
+	}
+}
+
+type retrievalMethod int
+
+const (
+	forSlot retrievalMethod = iota
+)
+
+// CanonicalBuilder builds a Replayer that uses a combination of database queries and a
+// CanonicalChecker (which should usually be a fork choice store implementing an IsCanonical method)
+// to determine the canonical chain and apply it to generate the desired state.
+type CanonicalBuilder struct {
+	chainer chainer
+}
+
+var _ ReplayerBuilder = &CanonicalBuilder{}
+
+func (r *CanonicalBuilder) ForSlot(target types.Slot) Replayer {
+	return &stateReplayer{chainer: r.chainer, method: forSlot, target: target}
+}
+
+// chainer is responsible for supplying the chain components necessary to rebuild a state,
+// namely a starting BeaconState and all available blocks from the starting state up to and including the target slot
+type chainer interface {
+	chainForSlot(ctx context.Context, target types.Slot) (state.BeaconState, []block.SignedBeaconBlock, error)
+}
+
+type canonicalChainer struct {
+	h  HistoryAccessor
+	c  CanonicalChecker
+	cs CurrentSlotter
+}
+
+var _ chainer = &canonicalChainer{}
+
+// ChainForSlot creates a value that satisfies the Replayer interface via db queries
+// and the stategen transition helper methods. This implementation uses the following algorithm:
+// - find the highest canonical block >= the target slot
+// - starting with this block, recursively search backwards for a stored state, and accumulate intervening blocks
+func (r *canonicalChainer) chainForSlot(ctx context.Context, target types.Slot) (state.BeaconState, []block.SignedBeaconBlock, error) {
+	ctx, span := trace.StartSpan(ctx, "canonicalChainer.ForSlot")
+	defer span.End()
+	currentSlot := r.cs.CurrentSlot()
+	if target > currentSlot {
+		return nil, nil, errors.Wrap(ErrFutureSlotRequested, fmt.Sprintf("requested=%d, current=%d", target, currentSlot))
+	}
+	_, b, err := r.canonicalBlockForSlot(ctx, target)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, fmt.Sprintf("unable to find replay data for slot=%d", target))
+	}
+	s, descendants, err := r.ancestorChain(ctx, b)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to query for ancestor and descendant blocks")
+	}
+
+	return s, descendants, nil
+}
+
+// canonicalBlockForSlot uses HighestSlotBlocksBelow(target+1) and the CanonicalChecker
+// to find the highest canonical block available to replay to the given slot.
+func (r *canonicalChainer) canonicalBlockForSlot(ctx context.Context, target types.Slot) ([32]byte, block.SignedBeaconBlock, error) {
+	for target > 0 {
+		hbs, err := r.h.HighestSlotBlocksBelow(ctx, target+1)
```

Comment:
This is out of scope of this PR, but why does the function return a slice when the implementation only ever returns one block? Even the doc says that it returns the highest **block** (singular).

Reply:
I agree, and have brought this up in discord a couple times. The reasoning is that the slot index is supposed to store multiple roots per slot, but in practice it does not. So I have written the code in anticipation of that being changed.